src/gluonts/mx/model/estimator.py

```diff
@@ -117,85 +122,98 @@ def create_predictor(
         """
         raise NotImplementedError
 
+    def create_training_data_loader(
+        self, data: Dataset, **kwargs
+    ) -> DataLoader:
+        raise NotImplementedError
+
+    def create_validation_data_loader(
+        self, data: Dataset, **kwargs
+    ) -> DataLoader:
+        raise NotImplementedError
+
     def train_model(
         self,
-        training_data: Dataset,
+        training_data: Optional[Dataset] = None,
         validation_data: Optional[Dataset] = None,
+        training_data_loader: Optional[DataLoader] = None,
+        validation_data_loader: Optional[DataLoader] = None,
         num_workers: Optional[int] = None,
         num_prefetch: Optional[int] = None,
         shuffle_buffer_length: Optional[int] = None,
-        **kwargs,
+        cache_data: bool = False,
     ) -> TrainOutput:
-        transformation = self.create_transformation()
-
-        # ensure that the training network is created within the same MXNet
-        # context as the one that will be used during training
-        with self.trainer.ctx:
-            trained_net = self.create_training_network()
-
-        input_names = get_hybrid_forward_input_names(trained_net)
-
-        training_data_loader = TrainDataLoader(
-            dataset=training_data,
-            transform=transformation + SelectFields(input_names),
-            batch_size=self.batch_size,
-            stack_fn=partial(
-                batchify,
-                ctx=self.trainer.ctx,
-                dtype=self.dtype,
-            ),
-            num_workers=num_workers,
-            num_prefetch=num_prefetch,
-            shuffle_buffer_length=shuffle_buffer_length,
-            decode_fn=partial(as_in_context, ctx=self.trainer.ctx),
-            **kwargs,
-        )
+        assert (training_data is None) != (
+            training_data_loader is None
+        ), "You must provide either `training_data` or `training_data_loader`"
```

Comment:
Can't we just do `training_data: Union[Dataset, DataLoader]`?

I don't like adding more arguments, if it's an either or thing.

Reply:
I like the idea, but I'm afraid python's typing wouldn't play well with that. How do you distinguish a `Dataset` from a `DataLoader`? They're essentially the same type `Iterable[Dict]`

Reply:
The alternative would be to have some other method `train_from_data_loaders` which is called by the current one

Reply:
Conclusion is: we can get rid of this. If someone wants to provide custom training/validation data loaders, they probably don't want to be using the "high level" estimator interface, and can just create the network and invoke

```python
trainer.train(net, training_data_loader, validation_data_loader)
```

It's on the single components to be usable as stand-alone, not on the estimator API to accommodate every possible usage.